### PR TITLE
Move transformationForTextBox() call out of the fragment-loop

### PR DESCRIPTION
**Pull request by Ahmad-S792**
#### e1696eb5587f9a3124b801f52328c8d36eff3077
<pre>
Move transformationForTextBox() call out of the fragment-loop

Move transformationForTextBox() call out of the fragment-loop
<a href="https://bugs.webkit.org/show_bug.cgi?id=250462">https://bugs.webkit.org/show_bug.cgi?id=250462</a>

Reviewed by Said Abou-Hallawa.

Merge - <a href="https://src.chromium.org/viewvc/blink?revision=177517&amp">https://src.chromium.org/viewvc/blink?revision=177517&amp</a>;view=revision

This patch is to refactor code since textBox is invariant, so there
is no need to fetch and check the transform for each fragment and
it is to refactor using modern C++ whole function.

Since this is just refactoring, it does not have any test case.

* Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
(SVGTextLayoutEngine::finalizeTransformMatrices): Refactor to not check transform for each fragment and use modern C++

Canonical link: <a href="https://commits.webkit.org/258886@main">https://commits.webkit.org/258886@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f6aa59e60c681c6787df306b4810bffe0458fa0c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103205 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12330 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/36206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112452 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172649 "Found 1 new test failure: js/dom/Promise-reject-large-string.html (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107158 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13355 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3234 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95428 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/110655 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108979 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10255 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/36206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37883 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92077 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/36206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79615 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5734 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/36206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5905 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2848 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11894 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/36206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6110 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7653 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by Ahmad-S792** [comment]
I am not sure on styling issue because it is still single statement on Line so adding {} makes no-sense. Any help would be appreciated!

File: Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp [hunk 1]
```diff
@@ -277,13 +279,14 @@ void SVGTextLayoutEngine::finalizeTransformMatrices(Vector<SVGInlineTextBox*>& b
...
         SVGInlineTextBox* textBox = boxes.at(boxPosition);
-        Vector<SVGTextFragment>& fragments = textBox->textFragments();
 
-        unsigned fragmentCount = fragments.size();
+    textBoxTransformation = m_chunkLayoutBuilder.transformationForTextBox(textBox);
+    if (textBoxTransformation.isIdentity())
+        continue;
+            
+    Vector<SVGTextFragment>& fragments = textBox->textFragments();
+    unsigned fragmentCount = fragments.size();
```
Comment (shallawa) [inline Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp]: You can use modern C++ for loop:

```
for (auto& fragment : textBox->textFragments()) {
    ASSERT(fragment.lengthAdjustTransform.isIdentity());
    fragment.lengthAdjustTransform = textBoxTransformation;
}

```

File: Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp [hunk 2]
```diff
@@ -277,13 +279,14 @@ void SVGTextLayoutEngine::finalizeTransformMatrices(Vector<SVGInlineTextBox*>& b
     AffineTransform textBoxTransformation;
     for (unsigned boxPosition = 0; boxPosition < boxCount; ++boxPosition) {
         SVGInlineTextBox* textBox = boxes.at(boxPosition);
-        Vector<SVGTextFragment>& fragments = textBox->textFragments();
 
-        unsigned fragmentCount = fragments.size();
+    textBoxTransformation = m_chunkLayoutBuilder.transformationForTextBox(textBox);
+    if (textBoxTransformation.isIdentity())
+        continue;
+
```
Comment (shallawa) [inline Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp]: Please fix the indentation and make sure they are spaces and not tabs. Also make sure the empty line after `continue` does not have any spaces. I think this is cause of the error you see from the style bot.

File: Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp [hunk 3]
```diff
@@ -277,13 +279,14 @@ void SVGTextLayoutEngine::finalizeTransformMatrices(Vector<SVGInlineTextBox*>& b
     AffineTransform textBoxTransformation;
     for (unsigned boxPosition = 0; boxPosition < boxCount; ++boxPosition) {
```
Comment (shallawa) [inline Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp]: You can make this also a modern C++ for loop:

```
void SVGTextLayoutEngine::finalizeTransformMatrices(Vector<SVGInlineTextBox*>& textBoxes)
{
    if (textBoxes.isEmpty())
        return;

    for (auto textBox : textBoxes) {
        auto textBoxTransformation = m_chunkLayoutBuilder.transformationForTextBox(textBox);
        if (textBoxTransformation.isIdentity())
            continue;

        for (auto& fragment : textBox->textFragments()) {
            ASSERT(fragment.lengthAdjustTransform.isIdentity());
            fragment.lengthAdjustTransform = textBoxTransformation;
        }
    }

    textBoxes.clear();
}
```

Comment (Ahmad-S792) [reply Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp]: Done.

File: Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp (hunk 2 above)
Comment (Ahmad-S792) [reply Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp]: Done.

File: Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp (hunk 1 above)
Comment (Ahmad-S792) [reply Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp]: Done. Thanks for help!

**Review by shallawa — APPROVED** [review]

**Pull request merged**